remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/RemoteMessagingConfigMatcher.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2021 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.remote.messaging.impl
+
+import com.duckduckgo.remote.messaging.api.RemoteMessage
+import com.duckduckgo.remote.messaging.impl.matchers.AndroidAppAttributeMatcher
+import com.duckduckgo.remote.messaging.impl.matchers.DeviceAttributeMatcher
+import com.duckduckgo.remote.messaging.impl.matchers.UserAttributeMatcher
+import com.duckduckgo.remote.messaging.impl.matchers.toResult
+import com.duckduckgo.remote.messaging.impl.models.MatchingAttribute
+import com.duckduckgo.remote.messaging.impl.models.RemoteConfig
+import com.duckduckgo.remote.messaging.impl.matchers.Result
+import timber.log.Timber
+
+class RemoteMessagingConfigMatcher(
+    val deviceAttributeMatcher: DeviceAttributeMatcher,
+    val androidAppAttributeMatcher: AndroidAppAttributeMatcher,
+    val userAttributeMatcher: UserAttributeMatcher
+) {
+    suspend fun evaluate(remoteConfig: RemoteConfig): RemoteMessage? {
+        val rules = remoteConfig.rules
+
+        remoteConfig.messages.forEach { message ->
+            val matchingRules = if (message.matchingRules.isEmpty() && message.exclusionRules.isEmpty()) return message else message.matchingRules
+
+            val matchingResult = matchingRules.evaluateMatchingRules(rules)
+            val excludeResult = message.exclusionRules.evaluateExclusionRules(rules)
+
+            if (matchingResult == Result.Match && excludeResult == Result.Fail) return message
+        }
+
+        return null
+    }
+
+    private suspend fun Iterable<Int>.evaluateMatchingRules(rules: Map<Int, List<MatchingAttribute>>): Result {
```

Comment:
we can remove the `suspend` (?)

Reply:
As discussed, this is affected by inner `suspend` coming from repositories